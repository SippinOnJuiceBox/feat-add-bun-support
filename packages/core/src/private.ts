/**
 * Utils used by the bundler when transforming code
 */

import type { EventsConsumer } from './events-consumer.js';
import type { QueueItem } from './global.js';
import type { Serializable } from './schemas.js';

export type StepFunction<
  Args extends Serializable[] = any[],
  Result extends Serializable | unknown = unknown,
> = ((...args: Args) => Promise<Result>) & {
  maxRetries?: number;
};

const registeredSteps = new Map<string, StepFunction>();

/**
 * Register a step function to be served in the server bundle
 */
export function registerStepFunction(stepId: string, stepFn: StepFunction) {
  registeredSteps.set(stepId, stepFn);
}

/**
 * Find a registered step function by name
 */
export function getStepFunction(stepId: string): StepFunction | undefined {
  return registeredSteps.get(stepId);
}

export interface WorkflowOrchestratorContext {
  globalThis: typeof globalThis;
  eventsConsumer: EventsConsumer;
  invocationsQueue: QueueItem[];
  onWorkflowError: (error: Error) => void;
  generateUlid: () => string;
  generateNanoid: () => string;
}
