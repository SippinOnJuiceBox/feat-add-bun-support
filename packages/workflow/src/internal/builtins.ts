/**
 * These are the built-in steps that are "automatically available" in the workflow scope. They are
 * similar to "stdlib" except that are not meant to be imported by users, but are instead "just available"
 * alongside user defined steps. They are used internally by the runtime
 */

export async function __builtin_response_array_buffer(res: Response) {
  'use step';
  return res.arrayBuffer();
}

export async function __builtin_response_json(res: Response) {
  'use step';
  return res.json();
}

export async function __builtin_response_text(res: Response) {
  'use step';
  return res.text();
}
