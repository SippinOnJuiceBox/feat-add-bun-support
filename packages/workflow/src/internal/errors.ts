export {
  WorkflowAPIError,
  WorkflowError,
  WorkflowRunCancelledError,
  WorkflowRunFailedError,
  WorkflowRunNotCompletedError,
  WorkflowRunNotFoundError,
  WorkflowRuntimeError,
} from '@workflow/errors';
