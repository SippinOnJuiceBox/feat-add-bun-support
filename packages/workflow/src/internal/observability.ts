/**
 * These are shared observability utilities that are used by the CLI and Web UIs.
 * They might later be moved to be part of the core API.
 */

export * from '@workflow/core/observability';
