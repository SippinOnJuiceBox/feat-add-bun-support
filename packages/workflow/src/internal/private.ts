/**
 * This is the "private" API of the workflow runtime that referenced by the compiler's output
 * but not meant to be document and used directly in user code.
 */

export * from '@workflow/core/private';
