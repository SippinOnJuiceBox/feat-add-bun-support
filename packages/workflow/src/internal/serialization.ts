export * from '@workflow/core/serialization';
