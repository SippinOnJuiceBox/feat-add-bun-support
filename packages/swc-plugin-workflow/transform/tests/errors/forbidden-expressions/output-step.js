import { registerStepFunction } from "workflow/internal/private";
/**__internal_workflows{"steps":{"input.js":{"stepWithArguments":{"stepId":"step//input.js//stepWithArguments"},"stepWithThis":{"stepId":"step//input.js//stepWithThis"}}}}*/;
export async function stepWithThis() {
    // Error: this is not allowed
    return this.value;
}
export async function stepWithArguments() {
    // Error: arguments is not allowed
    return arguments[0];
}
class TestClass extends BaseClass {
    async stepMethod() {
        'use step';
        // Error: super is not allowed
        return super.method();
    }
}
registerStepFunction("step//input.js//stepWithThis", stepWithThis);
registerStepFunction("step//input.js//stepWithArguments", stepWithArguments);
