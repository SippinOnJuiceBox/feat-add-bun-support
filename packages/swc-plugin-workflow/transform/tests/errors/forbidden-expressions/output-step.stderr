  x Instance methods cannot be marked with "use step". Only static methods, functions, and object methods are supported.
    ,-[input.js:14:1]
 13 |     class TestClass extends BaseClass {
 14 | ,->   async stepMethod() {
 15 | |       'use step';
 16 | |       // Error: super is not allowed
 17 | |       return super.method();
 18 | `->   }
 19 |     }
    `----
